src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1104 @@
+        return false;
+    }
+
+    // Does device mode allow modification of selected locations?
+    if (!mDelegate->IsSetSelectedLocationsAllowed(locationStatusText))
+    {
+        //        ChipLogError(Zcl, "AddSelectedLocation %u - %s", aSelectedLocation, locationStatusText);
+        return false;
+    }
+
+    return mDelegate->AddSelectedLocation(aSelectedLocation, dummyIndex);
+}
+
+bool Instance::ClearSelectedLocations()
+{
+    if (mDelegate->ClearSelectedLocations())
+    {
+        NotifySelectedLocationsChanged();
+        return true;
+    }
+
+    return false;
+}
+
+//*************************************************************************
+// Current Location manipulators
+
+DataModel::Nullable<uint32_t> Instance::GetCurrentLocation()
+{
+    return mCurrentLocation;
+}
+
+bool Instance::SetCurrentLocation(const DataModel::Nullable<uint32_t> & aCurrentLocation)
+{
+    // If not null, the value of this attribute SHALL match the LocationID field of an entry on the SupportedLocations attribute's
+    // list.
+    if ((!aCurrentLocation.IsNull()) && (!IsSupportedLocation(aCurrentLocation.Value())))
+    {
+        ChipLogError(Zcl, "SetCurrentLocation %u - location is not supported", aCurrentLocation.Value());
+        return false;
+    }
+
+    mCurrentLocation = aCurrentLocation;
+    NotifyCurrentLocationChanged();
+
+    // EstimatedEndTime SHALL be null if the CurrentLocation attribute is null.
+    if (mCurrentLocation.IsNull())
+    {
+        SetEstimatedEndTime(DataModel::Nullable<uint32_t>());
+    }
+
+    return true;
+}
+
+//*************************************************************************
+// Estimated End Time manipulators
+
+DataModel::Nullable<uint32_t> Instance::GetEstimatedEndTime()
+{
+    return mEstimatedEndTime;
+}
+
+bool Instance::SetEstimatedEndTime(const DataModel::Nullable<uint32_t> & aEstimatedEndTime)
+{
+    bool doChangeNotify = false;
+
+    // EstimatedEndTime SHALL be null if the CurrentLocation attribute is null.
+    if (mCurrentLocation.IsNull() && !aEstimatedEndTime.IsNull())
+    {
+        ChipLogError(Zcl, "SetEstimatedEndTime - must be null if Current Location is null");
+        return false;
+    }
+
+    // The value of this attribute SHALL only be reported in the following cases:
+    // - when it changes from 0 to any other value and vice versa
+    // - when it decreases
+    // - when it changes from null to any non-zero value
+    if ((!mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() == 0) &&
+         mEstimatedEndTime.Value() != 0) ||
+        (!mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() != 0) &&
+         mEstimatedEndTime.Value() == 0) ||
+        (!mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() < mEstimatedEndTime.Value())) ||
+        (mEstimatedEndTime.IsNull() && !aEstimatedEndTime.IsNull() && (aEstimatedEndTime.Value() != 0)))
```

Comment:
Hmm.  So changing from not-null-but-not-0 to null does not get reported?  Is that expected?

But assuming that's true (which seems odd), note that changing from null to any non-null value gets reported (either we are changing to 0, or we are changing to a nonzero value, both of which get reported) so this might be more readable as:
```suggestion
    if (mEstimatedEndTime == aEstimatedEndTime)
    {
        doChangeNotify = false;
    } 
    else if (mEstimatedEndTime.IsNull()) 
    {
        doChangeNotify = true;
    }
    else 
    {
        auto currentEndTime = mEstimatedEndTime.Value();
        if (currentEndTime == 0)
        {
            doChangeNotify = true;
        }
        else if (aEstimatedEndTime.IsNull())
        {
            doChangeNotify = false;
        } 
        else
        {
            auto newEndTime = aEstimatedEndTime.Value();
            // Note that here we know currentEndTime != newEndTime and currentEndTime > 0,
            // so the check below includes the "newEndTime is becoming 0" case automatically.
            doChangeNotify = (newEndTime < currentEndTime);
        }
    }
```

or some other sane way of breaking down the cases that can be edited as the spec changes without going slightly insane.

Reply:
100%

Reply:
move it to a method and then you can do early return (and not nest and else-if that deep)

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/0ed4491d9f753d06659d1c576058f87d5bc604ed).